MoreLinq/EquiZip.cs

```diff
@@ -168,6 +168,130 @@ static partial class MoreEnumerable
             return EquiZipImpl(first, second, third, fourth, resultSelector);
         }
 
+        /// <summary>
+        /// Returns tuples, where each tuple contains the N-th
+        /// element from each of the argument sequences. An exception is thrown
+        /// if the input sequences are of different lengths.
+        /// </summary>
+        /// <typeparam name="T1">Type of elements in first sequence</typeparam>
+        /// <typeparam name="T2">Type of elements in second sequence</typeparam>
+        /// <param name="first">The first sequence.</param>
+        /// <param name="second">The second sequence.</param>
+        /// <returns>
+        /// A sequence of tuples that contains elements of the two input sequences.
+        /// </returns>
+        /// <exception cref="InvalidOperationException">
+        /// The input sequences are of different lengths.
+        /// </exception>
+        /// <example>
+        /// <code><![CDATA[
+        /// var numbers = new[] { 1, 2, 3, 4 };
+        /// var letters = new[] { "A", "B", "C", "D" };
+        /// var zipped = numbers.EquiZip(letters);
+        /// ]]></code>
+        /// The <c>zipped</c> variable, when iterated over, will yield the tuples : (1, A),
+        /// (2, B), (3, C), (4, D) in turn.
+        /// </example>
+        /// <remarks>
+        /// This operator uses deferred execution and streams its results.
+        /// </remarks>
+
+        public static IEnumerable<(T1, T2)> EquiZip<T1, T2>(this IEnumerable<T1> first, IEnumerable<T2> second)
```

Comment:
I would simplify these overloads by simply delegating to the others that take a projection function:

```c#
        public static IEnumerable<(T1, T2)>
            EquiZip<T1, T2>(this IEnumerable<T1> first,
                            IEnumerable<T2> second) =>
                first.EquiZip(second, ValueTuple.Create);
```

Ditto for the others.

Reply:
Also, [`Zip` in .NET Core 3 has named the tuple elements as `First` and `Second`](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.zip?view=netcore-3.0#System_Linq_Enumerable_Zip__2_System_Collections_Generic_IEnumerable___0__System_Collections_Generic_IEnumerable___1__). I wonder if we should follow the same lead? My current thinking is that its somewhat pointless because `First` and `Second` is just about as arbitrary or good as the [tuple field names `Item1` and `Item2`](https://docs.microsoft.com/en-us/dotnet/api/system.valuetuple-2#fields). The only benefit of alignment in names would be that someone could replace `Zip` with `EquiZip` and then they wouldn't have to go changing the element names if they decided to use the give ones. Thoughts?

Reply:
I'm not fan of the delegating because in case of invalid argument, the exception will throw deeper.
I think the implem should be internal and friendly accessible from the test project and should be the target of the tests.
Consumer of the implem should also be tested (because it is responsible for throwing on invalid argument) and correct use of the implem.
No advice for the field names, and I don't see them on the link you provided.
I don't like 'first', 'second' so much, I prefer constant length field name.

Reply:
> I'm not fan of the delegating because in case of invalid argument, the exception will throw deeper.

I understand but the stack trace won't run deeper than a frame here so I think it's a reasonable compromise for sake of having to maintain less code and fewer tests. I would agree that a delegating overload should do its own argument checking if its arguments vary by name or some function but the tuple-returning overloads are mean to be merely dumb convenience wrappers.

> No strong opinions for the fields names, (I don't see them on the link you provided).

Documentation is currently should C++ usage but you can [check out the source and see them in there](https://github.com/dotnet/corefx/blob/v3.0.0/src/System.Linq/src/System/Linq/Zip.cs#L31):

```c#
public static IEnumerable<(TFirst First, TSecond Second)> Zip<TFirst, TSecond>(this IEnumerable<TFirst> first, IEnumerable<TSecond> second)
```

> No strong opinions for the fields names

Okay. This is something that can be reviewed separately.